### PR TITLE
[SPARK-22360][SQL][TEST] Add unit tests for Window Specifications

### PR DESCRIPTION
**Pull request by smurakozi**
## What changes were proposed in this pull request?

Improve the test coverage of window specifications.

New tests cover basic cases more systematically in DataFrameWindowFunctionsSuite:
- different partition clauses (none, one, multiple)
- different order clauses (none, one, multiple, asc/desc, nulls first/last)
- defaults if clauses are missing

New tests were added to cover some more complex cases when partitionBy or orderBy uses expressions.

ExpressionParserSuite.'window function expressions' was also extended to check parsing of some additional window expressions.

## How was this patch tested?
Only new tests were added, automated tests were executed.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 1]
```diff
@@ -246,13 +246,23 @@ class ExpressionParserSuite extends PlanTest {
...
     assertEqual("foo(*) over (distribute by a, b)", windowed(Seq('a, 'b)))
+    assertEqual("foo(*) over (cluster by a)", windowed(Seq('a)))
     assertEqual("foo(*) over (cluster by a, b)", windowed(Seq('a, 'b)))
-    assertEqual("foo(*) over (order by a desc, b asc)", windowed(Seq.empty, Seq('a.desc, 'b.asc )))
-    assertEqual("foo(*) over (sort by a desc, b asc)", windowed(Seq.empty, Seq('a.desc, 'b.asc )))
+    assertEqual("foo(*) over (order by a)", windowed(Seq.empty, Seq('a.asc)))
+    assertEqual("foo(*) over (order by a desc, b asc)", windowed(Seq.empty, Seq('a.desc, 'b.asc)))
+    assertEqual("foo(*) over (sort by a)", windowed(Seq.empty, Seq('a.asc)))
+    assertEqual("foo(*) over (sort by a desc, b asc)", windowed(Seq.empty, Seq('a.desc, 'b.asc)))
     assertEqual("foo(*) over (partition by a, b order by c)", windowed(Seq('a, 'b), Seq('c.asc)))
```
Comment (gaborgsomogyi) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: (partition by a order by c) is a possibility in order to be conform with the pattern you followed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 2]
```diff
@@ -246,13 +246,23 @@ class ExpressionParserSuite extends PlanTest {
...
     assertEqual("foo(*) over (distribute by a, b sort by c)", windowed(Seq('a, 'b), Seq('c.asc)))
```
Comment (gaborgsomogyi) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: (distribute by a order by c) is a possibility in order to be conform with the pattern you followed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala (hunk 1 above)
Comment (smurakozi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: I will fix it, thx

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala (hunk 2 above)
Comment (smurakozi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: I will fix it, thx

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala [hunk 3]
```diff
@@ -32,6 +32,217 @@ import org.apache.spark.unsafe.types.CalendarInterval
...
+
+  test("Multiple orderBy clauses with desc") {
+    val df = Seq(("a", "x", 1), ("a", "y", 2), ("b", "y", 3), ("b", "x", 4)).toDF("k1", "k2", "v")
+
+    checkAnswer(
+      df.select(
+        'v,
+        lead("v", 1).over(Window.orderBy($"k1".desc, $"k2")),
+        lead("v", 1).over(Window.orderBy($"k1", $"k2".desc)),
+        lead("v", 1).over(Window.orderBy($"k1".desc, $"k2"))),
```
[the remark this answers is not in the document]
Comment (smurakozi) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala]: It should be desc, desc, thanks for catching it.

**Comment by squito** [comment]
Jenkins, ok to test

**Comment by squito** [comment]
failure seems to be unrelated ...

Jenkins, retest this please

**Comment by smurakozi** [comment]
@gatorsmile @hvanhovell @jiangxb1987, could you have a look, please?

**Comment by smurakozi** [comment]
@gatorsmile @hvanhovell @jiangxb1987, could you have a look, please?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 4]
```diff
@@ -246,13 +246,24 @@ class ExpressionParserSuite extends PlanTest {
     // Basic window testing.
     assertEqual("foo(*) over w1", UnresolvedWindowExpression(func, WindowSpecReference("w1")))
     assertEqual("foo(*) over ()", windowed())
+    assertEqual("foo(*) over (partition by a)", windowed(Seq('a)))
```
Comment (jiangxb1987) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: You shall also cover the `null` cases, such as `foo(*) over (partition by null)`

**Comment by smurakozi** [comment]
Do you think I need to cover any other cases, @jiangxb1987 ?

**Comment by smurakozi** [comment]
I think I've addressed all your points @jiangxb1987. 
Do you think I need any more changes?

**Comment by jiangxb1987** [comment]
Sorry for the delay, I'll check the results later this week.

**Comment by smurakozi** [comment]
ping @jiangxb1987

**Comment by smurakozi** [comment]
ping @jiangxb1987

**Comment by rxin** [comment]
Can we add them to the file based test suites instead?

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
gentle ping @smurakozi

**Comment by dongjoon-hyun** [comment]
Do you still need this, @jiangxb1987 ? Also, gentle ping, @smurakozi .

**Comment by gaborgsomogyi** [comment]
I'm pretty sure @smurakozi will not finish this PR.

**Comment by HyukjinKwon** [comment]
Closing due to inactivity here.

**Pull request closed without merging**